source/features/useful-not-found-page.tsx

```diff
@@ -107,17 +114,53 @@ function init(): false | void {
 	for (let i = bar.children.length - 2; i >= 0; i--) {
 		void checkAnchor(bar.children[i] as HTMLAnchorElement);
 	}
+}
+
+async function showHelpfulLinks(): Promise<void> {
+	// What we should try to do to be helpful:
+	//  1. Check if the file is on the default branch
+	//  2. Check the current branch's history for the file (was it deleted or renamed)
+
+	const urlToFileOnDefaultBranch = await getUrlToFileOnDefaultBranch();
+	if (urlToFileOnDefaultBranch) {
+		select('h2')!.after(
+			<p className="container mt-4 text-center">
+				<a href={urlToFileOnDefaultBranch}>This {getType()}</a> exists on the default branch.
+			</p>,
+		);
+		return;
+	}
 
-	if (['tree', 'blob'].includes(parts[2])) {
-		// Object might be 410 Gone
-		void addCommitHistoryLink(bar);
+	const change = await getLatestChangeToFile();
+	if (change) {
+		const {fileInfo, commitInfo} = change;
+		const url = new GitHubURL(location.href);
+
+		url.assign({route: 'commits'});
+		const commitHistory = <a href={url.toString()}>Commit history</a>;
+		url.assign({route: 'blob', branch: commitInfo.parents[0].sha, filePath: url.filePath});
+		const lastVersion = <a href={fileInfo.status === 'removed' ? fileInfo.blob_url : url.toString()}>This {getType()}</a>;
+		const permalink = <a href={commitInfo.html_url}><relative-time datetime={commitInfo.commit.committer.date}/></a>;
+		const verb = fileInfo.status === 'removed'
+			? 'deleted'
+			: <a href={fileInfo.status === 'renamed' ? fileInfo.blob_url : url.toString()}>moved</a>;
+
+		select('h2')!.after(
+			<p className="container mt-4 text-center">
+				{lastVersion} was {verb} ({permalink}) - {commitHistory}.
+			</p>,
+		);
 	}
+}
 
-	if (['tree', 'blob', 'edit'].includes(parts[2])) {
-		// File might not be available on the current branch
-		// GitHub already redirects /tree/ and /blob/ natively
-		void addDefaultBranchLink(bar);
+function init(): false | void {
+	const parts = parseCurrentURL();
+	if (parts.length <= 1 || !select.exists('[alt*="This is not the web page you are looking for"]')) {
+		return false;
 	}
+
+	void showMissingPart();
+	void showHelpfulLinks();
```

Comment:
To speed things up, let's just split `showHelpfulLinks` into 2 unrelated functions. Even if they end up adding 2 lines, that's fine, it's extra information, it's all good. However we'll also need to **replace every `.after()` call** with something that `append`s  to an element so that we never risk moving existing elements. I think this works:

```js
select('main > .container-lg').before()
```

Reply:
Sure. Something like `showDefaultBranchLink` and `showAlternateFileLink`?

Reply:
Done!